### ISSUE TITLE
DeprecationWarning: asyncore (to be removed in Python 3.12)

### ISSUE THREAD

**Issue by Bastian-Krause** [lextudio/pysnmp#10]
This line..

https://github.com/lextudio/pysnmp/blob/002474a06f0383870ca28e73de6727eaa0d91431/pysnmp/carrier/asyncore/base.py#L9

..causes warnings such as:

```
[...]/python3.11/site-packages/pysnmp/carrier/asyncore/base.py:9: DeprecationWarning: The asyncore module is deprecated and will be removed in Python 3.12. The recommended replacement is asyncio
```
See also https://docs.python.org/3/library/asyncore.html. Python 3.12 will [probably be released in October 2023](https://peps.python.org/pep-0693/#schedule).

Btw, thanks for taking over this project!

**Comment by lextm** [comment]
Thanks for opening this. We do have a local version with such removed, but it has quite a lot of impact on documentation and so on, so we are now testing with a few clients and evaluate how to move on.

Will provide an update in a few weeks.

**Comment by Bastian-Krause** [comment]
Any update on this?

**Comment by Bastian-Krause** [comment]
@lextm The Python 3.12 release is expected in less than a month. It would be nice to have a pysnmp release with 3.12 support until then. Do you have any update on this?

**Comment by lextm** [comment]
Our internal Python 3.12 compliant build works for a few clients. However, they don't use every features of PySNMP, so not impacted by removal of the deprecated bits very much.

My team still works on porting as many examples as possible away from the deprecated API surface, but no promise right now on when we can finish that task.

More to disclose when Python 3.12 is released, and we don't aim to ship Python 3.12 on day 1. More likely to have a Python  3.12 target release by the end of this year.

**Comment by Bastian-Krause** [comment]
Thanks for sharing this info.

**Comment by lextm** [comment]
After internal discussion, our current plan has some slight changes. First, we will use a different version 6.0.x and make it solely available for Python 3.12. 

6.0.x release will provide a much smaller API surface since the underlying legacy APIs were removed.

Still aim to ship a stable release (6.0.3 or 6.0.4) by the end of 2023, but a few preview releases (from 6.0.0) will be published soon. The goal of the initial releases is to preview the API surface changes and collect feedback on the missing features and documentation.

**Comment by TimotheusBachinger** [comment]
Hi @lextm,
as we're approaching the end of the year, I just want to know if you could give us a short heads-up regarding the 6.0.x release.

We're at Checkmk are just about to lift Python to 3.12 and are heavily using pysnmp in our product :-)

Thanks for your great work!
Timi

**Comment by lextm** [comment]
@TimotheusBachinger Thanks for the advice.

We tested the workaround (by using pyasyncore) and it worked well. This buys us more time to work on the asyncore free 6.0 release.

Just shipped release 5.0.31 that contains the necessary changes, so please let us know if it works well for you.

**Comment by lextm** [comment]
We decided to completely dump asyncore bits in 6.0+ releases, and urge all 5.0.x users to ugprade.

The guide is https://docs.lextudio.com/pysnmp/upgrade
